### DISCUSSION TITLE
OrderedVocab error when using HuggingFace ConvBERT

### DISCUSSION THREAD

**Discussion by mbrunecky** [explosion/spaCy#11553] [Help: Coding & Implementations] [answered]
I am trying to train a transformer NER using HuggingFace ConvBERT model:

```
[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v3"
name = "YituTech/conv-bert-base"
```

After completing the initialization, I get approximately 1000 error messages like the following (index varies from about 5000 to 30519):

`The OrderedVocab you are attempting to save contains a hole for index 30519, your vocabulary could be corrupted
`

Following that, the training appears to proceed normally. Anybody can give me some hint as to where to look for 'OrderedVocab' and how serious is this message?

My data comes from OCR, hence it contains a lot of OOV 'words' (various garbage), so I am not surprised by excessive vocabulary. But I would like to understand what are the implications of this message...

**Comment by polm** [comment]
I would not expect OOV terms to cause an error like that.

I initially thought this was an error in HuggingFace Transformers somewhere, but there's no class named OrderedVocab there, or even your error string. I also [can't find it](https://github.com/yitu-opensource/ConvBert/search?q=orderedvocab) in the repo for the model you're using.

Does the error message have any further info with it, like a traceback or something? Or is it just a warning message with no context? If I search for your error message on Google I only get one hit, which a Chinese notebook using transformers, but a different model.

**Reply by mbrunecky** [reply]
Thank you, Paul, for looking into this:
There is no context for the message. And I stand corrected, it gets printed (times ~1000) on each model save. The training appears to converge, so the error seems to be just a 'warning'. I tried to find OrderedVocab myself (no luck),- may be it's part of the (dynamically downloaded) YituTech/conv-bert-base/pytorch_model.bin. 
 
PERHAPS there is some conflict between Spacy tokenization and ConvBERT tokenization. I am attaching my config file - just in case I am doing something blatantly wrong. Perhaps I need to specify transformer.model tokenizer_config = {"use_fast": true}, but that will be my next training run (at this point, I want to finish this run and evaluate the resulting model ... another half a day).

```
# This is a first attempt on using the ConvBERT (HuggingFace) transformer model
# batch_size = 128, transformer.model name = "YituTech/conv-bert-base"
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "en"
pipeline = ["transformer","ner","doc_cleaner"]
batch_size = 128
disabled = []
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[nlp.before_creation]
@callbacks = "adjust_stop_words"
add_stop_words = []
rem_stop_words = ["amount","and","as","at","between","by","eight","eleven","each","except","fifteen","fifty","first","five","for","formerly","forty","four","hereby","herein","nine","of","six","sixty","ten","third","three","to","twelve","twenty","two"]
debug = true

[components]

[components.doc_cleaner]
factory = "doc_cleaner"
silent = true

[components.doc_cleaner.attrs]
tensor = null
_.trf_data = null

[components.ner]
factory = "ner"
incorrect_spans_key = null
moves = null
scorer = {"@scorers":"spacy.ner_scorer.v1"}
update_with_oracle_cut_size = 128

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 80
maxout_pieces = 2
use_upper = false
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "*"

[components.transformer]
factory = "transformer"
max_batch_items = 2048
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v3"
name = "YituTech/conv-bert-base"
# perhaps? tokenizer_config = {"use_fast": true}
# mixed_precision = true

[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 80

[components.transformer.model.tokenizer_config]
use_fast = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 8000
max_epochs = 0
max_steps = 30400
eval_frequency = 800
frozen_components = []
before_to_disk = null
annotating_components = []

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 1536
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 30400
initial_rate = 0.00005

[training.score_weights]
ents_f = 0.5
ents_p = 0.2
ents_r = 0.3
ents_per_type = null

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

**Reply by polm** [reply]
I managed to track down the source of this warning, which is actually coming from Rust code in [Huggingface Tokenizers](https://github.com/huggingface/tokenizers/blob/5f6e9784526a4cd5e4f6dcdcc045cdceba5463e1/tokenizers/src/models/mod.rs#L42). It's not clear to me what would cause it, but I would carefully check your model outputs - if there's an issue, I would expect it to be immediately obvious.

> PERHAPS there is some conflict between Spacy tokenization and ConvBERT tokenization.

Differences in tokenization between spaCy and Transformers are normal, and there's actually already a significant amount of code dedicated to handling that specifically. However this issue is about a Huggingface Vocab object of some kind - it's different from the spaCy Vocab, but is still a mapping of tokens to IDs. A hole in the list of IDs is not something that would normally be caused simply by odd tokenization.

**Reply by mbrunecky** [reply]
Thank you a LOT!
As you properly suggested, the **saved** model outputs are bad. While evaluation during training (apparently using the in-memory-model) gives me reasonable p/r/f scores (~97%), running evaluation against the **saved** model gives p/r/f scores near zero.
Those 'holes' in the saved vocabulary probably mess everything up, even though (when using the saved model) I am not getting any errors.

Thanks again for the Huggingface link (I did not plan on looking at their code). I will try to figure out what could be causing the problem, and what (if anything) can I do about it. I opened a discussion item against Huggingface conv-bert-base, but it looks more like a generic Huggingface problem - perhaps one exposed by conv-bert-base.

**Reply by polm** [reply]
Glad you were able to catch the error! If you find out what the cause was, or it turns out there's something in spaCy that causes it, please do let us know.

**Comment by mbrunecky** [comment, marked as answer]
Thank you again, [polm](https://github.com/polm). Using your pointer, I was able to find other references to this problem, and develop a patch/workaround.
**It is not a Spacy problem**. The problem is in the original huggingface conv-bert-base contribution. I traced it to the 'vocab.txt' checked-in https://huggingface.co/YituTech/conv-bert-base. It contains (many) duplicates. There is 188 'tokens' replicated, many of them of them up to 34 times, for a total of 1008 'extra' (replicated) entries. Replicated entries are not loaded by the huggingface. And the replicated entries are (probably entirely) on Chinese tokens.

My 'solution' (workaround) is patching the 'vocab.txt' making the 'extra' (repeated) token entries unique by adding a suffix such as '00', '01'... resulting in '銆00' etc. Such tokens are highly unlikely in any text, especially in my case (English). It takes care of the 'contains hole' problem, and patched convert-bert-base now works with Spacy (3.4) as advertised. Early results suggest that it provides a bit better result for my NER domain, and the evaluation speed appears faster (not sure about training - need more testing).

I uploaded my patch as:
[vocab_patch.txt](https://github.com/huggingface/tokenizers/files/9699743/vocab_patch.txt)

and comments as:
https://github.com/huggingface/tokenizers/pull/954

**Reply by polm** [reply]
Thanks for the detailed report! Based on the linked issue, it looks like the base model did something wrong (by having duplicates in the vocab) and the Huggingface code handled it a little too optimistically. 

I'm not sure why duplicates would end up in a vocab like that. Since different characters have different numbers of repetitions, maybe it was linked to variation selectors or something...? These seem to be rare Chinese characters.

**Reply by mbrunecky** [reply]
The entire conv-bert-base contribution is over a year old. PERHAPS there was some problem (then) in either Huggingface or the transformer code (ConvBERT was not developed for Spacy). My training is adding a LOT to the vocabulary (it ends up having > 300,000 strings from the original ~35,000), and it does NOT create any 'gaps'. So I assume that _something_ happened at the time the original contribution was created. It would be very difficult to go back in time to reproduce the problem, so I give up on it. Besides '_it works for me_'  :-) now. 
Unfortunately, I am hopelessly behind on my project. And each Spacy upgrade (now 3.4) seems to pull the rug from under me.
